File: java/devtools/src/test/java/org/pactman/nonprofitcheckplus/devtools/ContractTest.java
package org.pactman.nonprofitcheckplus.devtools;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.pactman.nonprofitcheckplus.internal.Json;

/**
 * The response-signature engine the smoke and baseline tools share.
 *
 * <p>This is the part of the tooling with real logic in it, and the part whose
 * mistakes are silent: a differ that under-reports passes a green run over a
 * broken API, and one that over-reports gets switched off. The rules that are
 * easy to get subtly wrong are the ones about absence — a field missing because
 * it was removed, versus one missing because the object it lives in arrived
 * null. Getting that backwards either fails every green run or passes every
 * broken one, and neither is visible without a live deployment to try it
 * against, so it is pinned here.
 */
class ContractTest {

    private static Map<String, String> signature(String json) {
        return Contract.signatureOf(Json.parse(json));
    }

    private static Map<String, String> map(String... pairs) {
        Map<String, String> signature = new LinkedHashMap<>();

        for (int i = 0; i + 1 < pairs.length; i += 2) {
            signature.put(pairs[i], pairs[i + 1]);
        }

        return signature;
    }

    @ParameterizedTest
    @CsvSource({
        "411787097, digits:9",
        "01085-2643, digits:5-4",
        "00, digits:2",
        "'3/25/2026 3:28:54 PM', date",
        "'2026-08-24T09:47:53Z', date:iso",
        "https://pactman.org/profile, url",
        "'', empty",
        "'   ', empty",
        "EXAMPLE NONPROFIT, text",
    })
    @DisplayName("classifies a string by its form, never by its value")
    void classifiesStringsByForm(String value, String expected) {
        assertEquals(expected, Contract.formatOf(value));
    }

    @Test
    @DisplayName("both OFAC wordings share one token")
    void bothOfacWordingsShareOneToken() {
        assertEquals("ofac-sentence", Contract.formatOf(Fixtures.OFAC_NO_MATCH));
        assertEquals("ofac-sentence", Contract.formatOf(Fixtures.OFAC_POSSIBLE_MATCH));
    }

    @Test
    @DisplayName("a genuine wording change falls back to text")
    void wordingChangeFallsBackToText() {
        assertEquals(
                "text",
                Contract.formatOf("This organization is not on any watchlist we checked."));
    }

    @Test
    @DisplayName("flattens a response into paths and tokens")
    void flattensAResponse() {
        Map<String, String> signature = signature(
                "{\"code\":200,\"data\":{\"ein\":\"411787097\",\"pub78_verified\":true,"
                        + "\"revocation_code\":null,\"organization_types\":"
                        + "[{\"deductibility_limitation\":\"50%\"}]},\"errors\":null}");

        assertEquals("number", signature.get("code"));
        assertEquals("object", signature.get("data"));
        assertEquals("digits:9", signature.get("data.ein"));
        assertEquals("boolean", signature.get("data.pub78_verified"));
        assertEquals("null", signature.get("data.revocation_code"));
        assertEquals("array", signature.get("data.organization_types"));
        assertEquals("object", signature.get("data.organization_types[]"));
        assertEquals(
                "text", signature.get("data.organization_types[].deductibility_limitation"));
        assertEquals("null", signature.get("errors"));
    }

    @Test
    @DisplayName("every element of an array folds into one path")
    void arrayElementsFoldIntoOnePath() {
        Map<String, String> signature = signature(
                "{\"data\":[{\"ein\":\"411787097\"},{\"ein\":\"996589560\"}]}");

        assertEquals("array", signature.get("data"));
        assertEquals("object", signature.get("data[]"));
        assertEquals("digits:9", signature.get("data[].ein"));
        assertEquals(3, signature.size(), "two organizations describe one record shape");
    }

    @Test
    @DisplayName("a path with two forms records both, sorted")
    void mixedFormsAreRecordedTogether() {
        Map<String, String> signature = signature(
                "{\"data\":[{\"most_recent_bmf\":\"3/25/2026 3:28:54 PM\"},"
                        + "{\"most_recent_bmf\":null}]}");

        assertEquals("date|null", signature.get("data[].most_recent_bmf"));
    }

    @Test
    @DisplayName("records no value from the response")
    void recordsNoValues() {
        Map<String, String> signature = signature(
                "{\"data\":{\"ein\":\"411787097\",\"organization_name\":\"EXAMPLE NONPROFIT\"}}");

        for (String token : signature.values()) {
            assertFalse(token.contains("411787097"), "a signature must carry no value");
            assertFalse(token.contains("EXAMPLE"), "a signature must carry no value");
        }
    }

    @Test
    @DisplayName("schemaDiff reports fields that appeared and disappeared, and nothing else")
    void schemaDiffReportsPresenceOnly() {
        Difference diff = Contract.schemaDiff(
                map("kept", "text", "gone", "text", "moved", "text"),
                map("kept", "text", "moved", "number", "fresh", "text"));

        assertEquals(2, diff.total());
        assertEquals(Change.Kind.REMOVED, diff.changes().get(0).kind());
        assertEquals("gone", diff.changes().get(0).path());
        assertEquals(Change.Kind.ADDED, diff.changes().get(1).kind());
        assertEquals("fresh", diff.changes().get(1).path());
    }

    @Test
    @DisplayName("typeDiff reports only paths both signatures have")
    void typeDiffReportsSharedPathsOnly() {
        Difference diff = Contract.typeDiff(
                map("kept", "text", "gone", "text", "moved", "digits:9"),
                map("kept", "text", "moved", "text", "fresh", "text"));

        assertEquals(1, diff.total());
        assertEquals("moved", diff.changes().get(0).path());
        assertEquals("digits:9", diff.changes().get(0).from());
        assertEquals("text", diff.changes().get(0).to());
    }

    @Test
    @DisplayName("baselineDiff excuses a value that simply did not arrive this time")
    void baselineDiffExcusesNullability() {
        Difference diff = Contract.baselineDiff(
                map("data.pub78_city", "text", "data.most_recent_bmf", "date"),
                map("data.pub78_city", "null", "data.most_recent_bmf", "date|null"));

        assertTrue(diff.clean(), "neither difference is the API moving");
        assertEquals(2, diff.nullable());
    }

    @Test
    @DisplayName("baselineDiff still reports a move between two forms a value took")
    void baselineDiffReportsRealDrift() {
        Difference diff = Contract.baselineDiff(
                map("data.ein", "digits:9"), map("data.ein", "text"));

        assertEquals(1, diff.total());
        assertEquals("digits:9", diff.changes().get(0).from());
    }

    @Test
    @DisplayName("baselineDiff excuses paths under a parent that arrived null, in both directions")
    void baselineDiffExcusesUnreachablePaths() {
        Map<String, String> populated = map(
                "data.organization_types", "array",
                "data.organization_types[]", "object",
                "data.organization_types[].deductibility_limitation", "text");
        Map<String, String> empty = map("data.organization_types", "null");

        Difference forwards = Contract.baselineDiff(populated, empty);
        Difference backwards = Contract.baselineDiff(empty, populated);

        assertTrue(forwards.clean());
        assertEquals(2, forwards.unreachable());
        assertTrue(backwards.clean());
        assertEquals(2, backwards.unreachable());
    }

    @Test
    @DisplayName("an empty array leaves its members unreachable too")
    void emptyArrayLeavesMembersUnreachable() {
        Difference diff = Contract.baselineDiff(
                map("errors", "array", "errors[]", "object", "errors[].reason", "text"),
                map("errors", "array"));

        assertTrue(diff.clean());
        assertEquals(2, diff.unreachable());
    }

    @Test
    @DisplayName("permits treats `string` as a wildcard over every string form")
    void stringIsAWildcard() {
        assertTrue(Contract.permits("null|string", "text"));
        assertTrue(Contract.permits("null|string", "date"));
        assertTrue(Contract.permits("null|string", "digits:9"));
        assertTrue(Contract.permits("null|string", "url"));
        assertTrue(Contract.permits("null|string", "null"));

        // Where the contract names a form, a different string form fails.
        assertTrue(Contract.permits("digits:9|null", "digits:9"));
        assertFalse(Contract.permits("digits:9|null", "text"));
        assertFalse(Contract.permits("digits:9|null", "digits:5-4"));

        assertFalse(Contract.permits("null|string", "number"));
        assertFalse(Contract.permits("boolean|null", "text"));
    }

    @Test
    @DisplayName("composeExpected describes one record shape for both endpoints")
    void composeExpectedSharesOneRecordShape() {
        Map<String, Object> contract = ContractResources.contract();
        Map<String, String> single = Contract.composeExpected(contract, Contract.Kind.SINGLE);
        Map<String, String> bulk = Contract.composeExpected(contract, Contract.Kind.BULK);

        assertEquals("null|object", single.get("data"));
        assertEquals("array|null", bulk.get("data"));
        assertEquals("object", bulk.get("data[]"));

        // The same field, described identically under each prefix.
        assertEquals(single.get("data.ein"), bulk.get("data[].ein"));
        assertEquals(
                single.get("data.organization_types[].deductibility_limitation"),
                bulk.get("data[].organization_types[].deductibility_limitation"));
        assertEquals("null|object", single.get("data.organization_types[]"));
    }

    @Test
    @DisplayName("coverageDiff passes a response whose absent paths sit under a null parent")
    void coverageDiffExcusesUnreachablePaths() {
        Map<String, String> expected = map(
                "code", "number",
                "data", "null|object",
                "data.ein", "digits:9",
                "data.organization_types", "array|null",
                "data.organization_types[]", "null|object",
                "data.organization_types[].organization_type", "null|string",
                "errors", "array|null|string",
                "errors[]", "object",
                "errors[].reason", "string");

        Map<String, String> observed = map(
                "code", "number",
                "data", "object",
                "data.ein", "digits:9",
                "data.organization_types", "null",
                "errors", "null");

        Difference diff = Contract.coverageDiff(expected, observed, null);

        assertTrue(diff.clean());
        // errors[], errors[].reason, organization_types[] and its one field.
        assertEquals(4, diff.unreachable());
    }

    @Test
    @DisplayName("coverageDiff reports a vanished container once, at its shallowest path")
    void coverageDiffReportsAContainerOnce() {
        Map<String, String> expected = map(
                "code", "number",
                "data", "null|object",
                "data.ein", "digits:9",
                "data.organization_name", "null|string");

        Difference diff = Contract.coverageDiff(expected, map("code", "number"), null);

        assertEquals(1, diff.total());
        assertEquals("data", diff.changes().get(0).path());
    }

    @Test
    @DisplayName("coverageDiff reports a field the API invented")
    void coverageDiffReportsUnpredictedFields() {
        Difference diff = Contract.coverageDiff(
                map("code", "number"), map("code", "number", "data.brand_new", "text"), null);

        assertEquals(1, diff.total());
        assertEquals(Change.Kind.ADDED, diff.changes().get(0).kind());
        assertEquals("data.brand_new", diff.changes().get(0).path());
    }

    @Test
    @DisplayName("coverageDiff excuses an absent field nothing promises is present")
    void coverageDiffExcusesOptionalAbsence() {
        Map<String, String> expected =
                map("code", "number", "data", "object", "data.address_line2", "null|string");
        Map<String, String> observed = map("code", "number", "data", "object");

        Difference required = Contract.coverageDiff(expected, observed, null);
        Difference optional = Contract.coverageDiff(
                expected, observed, Set.of("code", "data"));

        assertEquals(1, required.total(), "with no policy, every predicted path is required");
        assertTrue(optional.clean());
        assertEquals(1, optional.optionalAbsent());
    }

    @Test
    @DisplayName("this SDK promises no field is always present, so nothing is required of a record")
    void nothingIsRequiredOfARecord() {
        Map<String, Object> contract = ContractResources.contract();
        Set<String> required = Contract.requiredPathsOf(contract, Contract.Kind.SINGLE);

        // Only the structural paths the envelope always has.
        assertEquals(
                Set.of("data", "errors[]", "errors[].eins[]", "data.organization_types[]"),
                required);
    }

    @Test
    @DisplayName("removals are reported before changes, and changes before additions")
    void changesAreOrderedBySeverity() {
        Difference diff = Contract.baselineDiff(
                map("gone", "text", "moved", "digits:9"),
                map("moved", "text", "fresh", "text"));

        assertEquals(
                Arrays.asList(Change.Kind.REMOVED, Change.Kind.CHANGED, Change.Kind.ADDED),
                Arrays.asList(
                        diff.changes().get(0).kind(),
                        diff.changes().get(1).kind(),
                        diff.changes().get(2).kind()));
    }

    @Test
    @DisplayName("summarize names only the counts that are not zero")
    void summarizeNamesNonZeroCounts() {
        Difference diff = Contract.schemaDiff(map("gone", "text"), map("fresh", "text"));

        assertEquals("1 removed, 1 added", Contract.summarize(diff.changes()));
        assertEquals("no differences", Contract.summarize(java.util.Collections.emptyList()));
    }

    @Test
    @DisplayName("the fixture API answers with a shape the contract permits")
    void theFixtureApiMatchesTheContract() throws Exception {
        // The stand-in API is what every example and the smoke runner exercise.
        // If it drifted from the contract, all of them would be rehearsing
        // against a shape the real service does not produce.
        Map<String, Object> contract = ContractResources.contract();
        Map<String, String> expected =
                Contract.composeExpected(contract, Contract.Kind.SINGLE);

        try (FixtureApi api = FixtureApi.start("test-key")) {
            java.net.http.HttpClient client = java.net.http.HttpClient.newHttpClient();
            java.net.http.HttpRequest request = java.net.http.HttpRequest.newBuilder(
                            java.net.URI.create(api.baseUrl()
                                    + "/api/entities/nonprofitcheck/v1/us/ein/"
                                    + FixtureEins.PUBLIC_CHARITY))
                    .header("Authorization", "Bearer test-key")
                    .build();

            String body = client.send(
                            request, java.net.http.HttpResponse.BodyHandlers.ofString())
                    .body();

            Map<String, String> observed = Contract.signatureOf(Json.parse(body));
            Difference types = Contract.contractDiff(expected, observed);
            Difference coverage = Contract.coverageDiff(
                    expected, observed, Contract.requiredPathsOf(contract, Contract.Kind.SINGLE));

            assertTrue(
                    types.clean(),
                    "the fixture API carries a value the contract forbids: "
                            + Contract.formatChanges(types.changes(), ""));
            assertTrue(
                    coverage.clean(),
                    "the fixture API and the contract disagree about fields: "
                            + Contract.formatChanges(coverage.changes(), ""));
        }
    }
}

File: go/internal/contract/contract_test.go
package contract

import (
	"bytes"
	"encoding/json"
	"os"
	"reflect"
	"slices"
	"sort"
	"strings"
	"testing"

	"github.com/PledgeSoftwareTX/new-pactman-nonprofitcheck-api-sdks/go/pactman"
)

// The smallest expectation that still has a nested object and an array in it.
var expected = Signature{
	"code":                      "number",
	"data":                      "null|object",
	"data.ein":                  "digits:9",
	"data.organization_types":   "array|null",
	"data.organization_types[]": "object",
	"data.organization_types[].organization_type": "null|string",
	"errors":          "array|null|string",
	"errors[]":        "object",
	"errors[].reason": "string",
}

// A successful response: no errors, and this organization has no types.
var success = Signature{
	"code":                    "number",
	"data":                    "object",
	"data.ein":                "digits:9",
	"data.organization_types": "null",
	"errors":                  "null",
}

func with(base Signature, changes map[string]string) Signature {
	out := Signature{}

	for path, token := range base {
		out[path] = token
	}

	for path, token := range changes {
		if token == "" {
			delete(out, path)
		} else {
			out[path] = token
		}
	}

	return out
}

func assertChanges(t *testing.T, got Diff, want ...Change) {
	t.Helper()

	if want == nil {
		want = []Change{}
	}

	if !reflect.DeepEqual(got.Changes, want) || got.Total != len(want) {
		t.Errorf("changes = %+v, want %+v", got.Changes, want)
	}
}

func TestCoverageDiff(t *testing.T) {
	t.Run("passes a response whose absent paths all sit under a null parent", func(t *testing.T) {
		result := CoverageDiff(expected, success, nil)

		assertChanges(t, result)

		// errors[], errors[].reason, organization_types[] and its one field.
		if result.Unreachable != 4 {
			t.Errorf("unreachable = %d", result.Unreachable)
		}
	})

	t.Run("fails a field that went missing while its parent was there", func(t *testing.T) {
		assertChanges(t, CoverageDiff(expected, with(success, map[string]string{"data.ein": ""}), nil),
			Change{Kind: "removed", Path: "data.ein", Token: "digits:9"})
	})

	t.Run("fails a field the API invented", func(t *testing.T) {
		assertChanges(t, CoverageDiff(expected, with(success, map[string]string{"data.new_field": "text"}), nil),
			Change{Kind: "added", Path: "data.new_field", Token: "text"})
	})

	t.Run("reports a container that vanished once, not once per field under it", func(t *testing.T) {
		assertChanges(t, CoverageDiff(expected, Signature{"code": "number", "errors": "null"}, nil),
			Change{Kind: "removed", Path: "data", Token: "null|object"})
	})

	t.Run("treats an array that arrived empty as having no room for its elements", func(t *testing.T) {
		assertChanges(t, CoverageDiff(expected, with(success, map[string]string{"data.organization_types": "array"}), nil))
	})

	t.Run("fails a field missing from the elements an array did return", func(t *testing.T) {
		observed := with(success, map[string]string{"data.organization_types": "array", "data.organization_types[]": "object"})

		assertChanges(t, CoverageDiff(expected, observed, nil),
			Change{Kind: "removed", Path: "data.organization_types[].organization_type", Token: "null|string"})
	})

	t.Run("counts an absent optional field instead of failing it", func(t *testing.T) {
		result := CoverageDiff(expected, with(success, map[string]string{"data.ein": ""}), map[string]bool{"data": true})

		assertChanges(t, result)

		if result.OptionalAbsent != 1 {
			t.Errorf("optional absent = %d", result.OptionalAbsent)
		}
	})
}

func TestRecordedBaseline(t *testing.T) {
	recorded := Signature{"code": "number", "data.ein": "digits:9", "data.city": "text"}

	t.Run("reports a path that appeared and a path that disappeared", func(t *testing.T) {
		now := Signature{"code": "number", "data.ein": "digits:9", "data.county": "text"}

		assertChanges(t, SchemaDiff(recorded, now),
			Change{Kind: "removed", Path: "data.city", Token: "text"},
			Change{Kind: "added", Path: "data.county", Token: "text"})
	})

	t.Run("reports a value whose form moved, on a path both have", func(t *testing.T) {
		assertChanges(t, TypeDiff(recorded, with(recorded, map[string]string{"data.ein": "digits:2-7"})),
			Change{Kind: "changed", Path: "data.ein", From: "digits:9", To: "digits:2-7"})
	})

	t.Run("sees no difference in an identical signature", func(t *testing.T) {
		if SchemaDiff(recorded, with(recorded, nil)).Total != 0 || TypeDiff(recorded, with(recorded, nil)).Total != 0 {
			t.Error("an identical signature differed")
		}
	})
}

func TestBaselineDiff(t *testing.T) {
	// One organization, recorded on an afternoon its Pub 78 row was populated.
	recorded := Signature{
		"code":                      "number",
		"data.ein":                  "digits:9",
		"data.pub78_city":           "text",
		"data.organization_types":   "array",
		"data.organization_types[]": "object",
		"data.organization_types[].organization_type": "text",
		"errors": "null",
	}

	t.Run("passes a subject whose nullable fields came back empty this time", func(t *testing.T) {
		now := Signature{
			"code":                    "number",
			"data.ein":                "digits:9",
			"data.pub78_city":         "null",
			"data.organization_types": "null",
			"errors":                  "null",
		}

		result := BaselineDiff(recorded, now)

		assertChanges(t, result)

		if result.Nullable != 2 || result.Unreachable != 2 {
			t.Errorf("nullable %d, unreachable %d", result.Nullable, result.Unreachable)
		}
	})

	t.Run("passes a field that filled in since the recording", func(t *testing.T) {
		result := BaselineDiff(Signature{"data.pub78_city": "null"}, Signature{"data.pub78_city": "text"})

		assertChanges(t, result)

		if result.Nullable != 1 {
			t.Errorf("nullable = %d", result.Nullable)
		}
	})

	t.Run("passes a token that only gained or lost null", func(t *testing.T) {
		result := BaselineDiff(Signature{"data[].address_line2": "digits:4|null"}, Signature{"data[].address_line2": "digits:4"})

		assertChanges(t, result)
	})

	t.Run("passes the paths under a parent that was null when it was recorded", func(t *testing.T) {
		result := BaselineDiff(Signature{"errors": "null"},
			Signature{"errors": "array", "errors[]": "object", "errors[].code": "number"})

		assertChanges(t, result)

		if result.Unreachable != 2 {
			t.Errorf("unreachable = %d", result.Unreachable)
		}
	})

	t.Run("still fails a value whose form moved between two real forms", func(t *testing.T) {
		result := BaselineDiff(Signature{"data.ein": "digits:9"}, Signature{"data.ein": "text"})

		assertChanges(t, result, Change{Kind: "changed", Path: "data.ein", From: "digits:9", To: "text"})
	})

	t.Run("still fails a form that moved while the field also turned nullable", func(t *testing.T) {
		assertChanges(t, BaselineDiff(Signature{"data.ein": "digits:9"}, Signature{"data.ein": "null|text"}),
			Change{Kind: "changed", Path: "data.ein", From: "digits:9", To: "null|text"})
	})

	t.Run("still fails a field that vanished while its parent was there", func(t *testing.T) {
		assertChanges(t, BaselineDiff(recorded, with(recorded, map[string]string{"data.ein": ""})),
			Change{Kind: "removed", Path: "data.ein", Token: "digits:9"})
	})

	t.Run("still fails a field the API invented", func(t *testing.T) {
		assertChanges(t, BaselineDiff(recorded, with(recorded, map[string]string{"data.county": "text"})),
			Change{Kind: "added", Path: "data.county", Token: "text"})
	})
}

func TestComposeExpected(t *testing.T) {
	contract := Contract{
		Envelope:         map[string]string{"code": "number", "errors": "array|null|string"},
		ErrorDetail:      map[string]string{"reason": "string"},
		Nonprofit:        map[string]string{"ein": "digits:9|null"},
		OrganizationType: map[string]string{"organization_type": "null|string"},
	}

	single, bulk := ComposeExpected(contract, "single"), ComposeExpected(contract, "bulk")

	if single["data.ein"] != "digits:9|null" || bulk["data[].ein"] != "digits:9|null" {
		t.Errorf("record paths: %q, %q", single["data.ein"], bulk["data[].ein"])
	}

	if single["data"] != "null|object" || bulk["data"] != "array|null" {
		t.Errorf("data: %q, %q", single["data"], bulk["data"])
	}

	if single["data.organization_types[]"] != "null|object" || bulk["data[].organization_types[]"] != "null|object" {
		t.Error("an element of organization_types should be allowed to be null")
	}
}

func TestSignatureOf(t *testing.T) {
	body := []byte(`{
		"code": 200, "errors": null,
		"data": [
			{"ein": "411787097", "zip": "01085-2643", "most_recent_bmf": "12/09/2025 12:00:00 AM",
			 "organization_types": [{"deductibility_limitation": "50%"}]},
			{"ein": "996589560", "zip": null, "most_recent_bmf": "12/09/2025 12:00:00 AM",
			 "organization_types": []}
		]
	}`)

	signature, err := SignatureOf(body)
	if err != nil {
		t.Fatal(err)
	}

	want := Signature{
		"code":                        "number",
		"errors":                      "null",
		"data":                        "array",
		"data[]":                      "object",
		"data[].ein":                  "digits:9",
		"data[].zip":                  "digits:5-4|null",
		"data[].most_recent_bmf":      "date",
		"data[].organization_types":   "array",
		"data[].organization_types[]": "object",
		"data[].organization_types[].deductibility_limitation": "text",
	}

	if !reflect.DeepEqual(signature, want) {
		t.Errorf("signature = %v", signature)
	}
}

func TestFormatOf(t *testing.T) {
	cases := map[string]string{
		"411787097":              "digits:9",
		"01085-2643":             "digits:5-4",
		"00":                     "digits:2",
		"12/09/2025 12:00:00 AM": "date",
		"3/7/2026, 9:41:03 PM":   "date",
		"2025-12-09T00:00:00Z":   "date:iso",
		"https://pactman.org/x":  "url",
		"":                       "empty",
		"   ":                    "empty",
		"EXAMPLE NONPROFIT":      "text",
		"This organization was NOT included in the Office of Foreign Assets Control Specially Designated Nationals (SDN) list.":                                       "ofac-sentence",
		"This organization may be included in the Office of Foreign Assets Control Specially Designated Nationals(SDN) list. A close match was found with UID: 41234": "ofac-sentence",
	}

	for value, want := range cases {
		if got := FormatOf(value); got != want {
			t.Errorf("FormatOf(%q) = %s, want %s", value, got, want)
		}
	}
}

// --- the contract against the Node SDK and against the Go models ----------

func TestContractIsTheNodeSDKsCopy(t *testing.T) {
	node, err := os.ReadFile("../../../nodejs/src/response-contract.json")
	if err != nil {
		t.Skipf("the Node SDK is not checked out beside this one: %v", err)
	}

	if !bytes.Equal(node, contractJSON) {
		t.Fatal("internal/contract/response-contract.json has drifted from nodejs/src/response-contract.json; " +
			"copy the Node file over it — the Node SDK is the source of truth for the contract")
	}
}

// shapes pairs each contract section with the model that reads it.
var shapes = []struct {
	section string
	model   reflect.Type
}{
	{"nonprofit", reflect.TypeOf(pactman.Nonprofit{})},
	{"envelope", reflect.TypeOf(pactman.Envelope{})},
	{"errorDetail", reflect.TypeOf(pactman.APIErrorDetail{})},
	{"organizationType", reflect.TypeOf(pactman.OrganizationType{})},
}

func section(t *testing.T, name string) map[string]string {
	t.Helper()

	contract, err := Load()
	if err != nil {
		t.Fatal(err)
	}

	return map[string]map[string]string{
		"nonprofit":        contract.Nonprofit,
		"envelope":         contract.Envelope,
		"errorDetail":      contract.ErrorDetail,
		"organizationType": contract.OrganizationType,
	}[name]
}

// tagged maps each json tag on a model to its field index.
func tagged(model reflect.Type) map[string]int {
	fields := map[string]int{}

	for i := 0; i < model.NumField(); i++ {
		name, _, _ := strings.Cut(model.Field(i).Tag.Get("json"), ",")
		if name != "" && name != "-" {
			fields[name] = i
		}
	}

	return fields
}

func keys[V any](m map[string]V) []string {
	out := make([]string, 0, len(m))
	for key := range m {
		out = append(out, key)
	}

	sort.Strings(out)

	return out
}

func TestModelsDeclareExactlyTheContractFields(t *testing.T) {
	for _, shape := range shapes {
		if got, want := keys(tagged(shape.model)), keys(section(t, shape.section)); !slices.Equal(got, want) {
			t.Errorf("%s declares %v\ncontract %s predicts %v", shape.model.Name(), got, shape.section, want)
		}
	}
}

// sentinelFor is a JSON value of the kind a token predicts. A shape, never a
// realistic value: a test fed real EINs would pass for the wrong reason the
// day a decoder began parsing them. The list holds an object and a string so
// it satisfies both the record list and the EIN list.
func sentinelFor(token string) string {
	switch token {
	case "null":
		return "null"
	case "boolean":
		return "true"
	case "number":
		return "1"
	case "array":
		return `[{"organization_type":"sentinel"},"sentinel"]`
	case "object":
		return `{"field":"sentinel"}`
	case "empty":
		return `""`
	}

	return `"sentinel"`
}

func TestEveryPredictedShapeIsReadable(t *testing.T) {
	for _, shape := range shapes {
		fields := tagged(shape.model)

		for field, tokens := range section(t, shape.section) {
			for _, token := range strings.Split(tokens, "|") {
				body := []byte(`{"` + field + `":` + sentinelFor(token) + `}`)
				decoded := reflect.New(shape.model)

				if err := json.Unmarshal(body, decoded.Interface()); err != nil {
					t.Fatalf("%s: %v", body, err)
				}

				value := decoded.Elem().Field(fields[field])
				wire := decoded.Elem().FieldByName("Fields").Interface().(pactman.Object)

				if token == "null" {
					if !wire.Has(field) || !value.IsZero() {
						t.Errorf("%s.%s returned as null should be present in Fields and nil on the model", shape.section, field)
					}

					continue
				}

				if value.IsZero() || (value.Kind() == reflect.Slice && value.Len() == 0) {
					t.Errorf("%s.%s predicts a %s value, but %s.%s reads nothing from one",
						shape.section, field, token, shape.model.Name(), shape.model.Field(fields[field]).Name)
				}
			}
		}
	}
}

func TestContractVocabulary(t *testing.T) {
	known := map[string]bool{"null": true, "boolean": true, "number": true, "array": true, "object": true, "string": true}

	contract, err := Load()
	if err != nil {
		t.Fatal(err)
	}

	for _, shape := range shapes {
		for field, tokens := range section(t, shape.section) {
			for _, token := range strings.Split(tokens, "|") {
				if !known[token] && !IsStringToken(token) {
					t.Errorf("%s.%s uses unknown token %q", shape.section, field, token)
				}
			}

			if strings.Contains(tokens, "411787097") || strings.Contains(tokens, "EXAMPLE") {
				t.Errorf("%s.%s looks like a value, not a shape", shape.section, field)
			}
		}

		// Every model field is optional, so this package makes no presence
		// promise at all; an entry here would be one the types cannot keep.
		required, ok := contract.Required[shape.section]
		if !ok || len(required) != 0 {
			t.Errorf("required.%s = %v, want an empty list", shape.section, required)
		}
	}
}

func TestEmbeddedBaselineReads(t *testing.T) {
	baseline, err := LoadBaseline()
	if err != nil || baseline.Single == nil || baseline.Bulk == nil || len(baseline.Single.Signature) == 0 {
		t.Fatalf("baseline = %+v, %v", baseline, err)
	}
}
